File: src/main/java/dev/akif/cookbook/recipe/RecipeRepository.java
package dev.akif.cookbook.recipe;

import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.ListCrudRepository;
import org.springframework.data.repository.PagingAndSortingRepository;
import org.springframework.stereotype.Repository;

import java.util.List;

@Repository
public interface RecipeRepository extends PagingAndSortingRepository<RecipeEntity, Long>, ListCrudRepository<RecipeEntity, Long> {
    @Query("""
           SELECT r
           FROM recipe r
           WHERE (:name IS NULL OR LOWER(r.name) LIKE CONCAT("%", LOWER(:name), "%")) AND (:numberOfServings IS NULL OR r.numberOfServings >= :numberOfServings)
           ORDER BY r.name, r.numberOfServings
           """)
    List<RecipeEntity> findAllBy(String name, Integer numberOfServings);
}

File: src/main/kotlin/dev/akif/cookbook/recipe/RecipeRepository.kt
package dev.akif.cookbook.recipe

import kotlinx.coroutines.flow.Flow
import org.springframework.data.r2dbc.repository.Query
import org.springframework.data.repository.kotlin.CoroutineCrudRepository
import org.springframework.stereotype.Repository

@Repository
interface RecipeRepository: CoroutineCrudRepository<RecipeEntity, Long> {
    @Query(
        """SELECT r.id, r.name, r.number_of_servings
           FROM recipes r
           WHERE (:name IS NULL OR LOWER(r.name) LIKE CONCAT('%', LOWER(:name), '%')) AND (:numberOfServings IS NULL OR r.number_of_servings >= :numberOfServings)
           ORDER BY r.name, r.number_of_servings"""
    )
    fun findAllBy(name: String?, numberOfServings: Int?): Flow<RecipeEntity>
}
